<div style="border:solid lightblue 3px; padding: 20px">
<b>Михаил, привет!</b>

Меня зовут Виталий Мехоношин, и я буду проверять твой проект. Предлагаю общаться на «ты» 😉

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Тобой проделана большая работа над проектом, давай сделаем его еще лучше. Ниже ты найдешь мои комментарии. Есть несколько правил:<br>
- <b>Пожалуйста, не перемещай, не изменяй и не удаляй комментарии.</b>
- У комментариев могут быть приписки <b>V2, V3</b> и т.д., которые соответствуют версии проверки. 
- Еще обрати внимание на цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b> 
Комментарии такого цвета необходимо исправить, без их отработки проект не будет принят. 
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b> 
На комментарии такого цвета нужно обязательно обратить внимание. Их не должно быть много.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> 
Комментарии такого цвета будут содержать лайфхаки, которые смогут улучшить твой верный код, их необязательно исправлять.
</div>

Предлагаю работать над проектом в диалоге: если ты хочешь ответить на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии вот так:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее.

<div style="border:solid lightblue 3px; padding: 20px">
<b>Общий комментарий ревьюера:</b>

У тебя отличная работа, но стоит поработать над следующими моментами:
    
1. Поработай над заполнением пропусков в `days_employed` и `total_income`.
2. Поработай над предобработкой нулевых значений в столбце с возрастом.
3. Выводи для начала частотность слов после лемматизации, чтобы твои категории явно выделились, для этого они должны быть в топе, иначе непонятно откуда они взялись.
4. Поработай над расчетом долей при ответе на вопрос о влиянии дохода на должность.
    
Соответствующик комменты найдешь в блокноте. Буду ждать твоих исправлений😊

<div style="border:solid lightblue 3px; padding: 20px">
<b>Общий комментарий ревьюера V2:</b>

Работа по комментариям успешно произведена, проект принят, удачи на следующих спринтах😊

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Старайся расширять стандартное описание к проекту.
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> А как, например, не совсем понял, это же техническое задание, для чего его расширять?</div>

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

По дефолту в блокноте дано краткое описание. В курсе же дано полное описание проекта с расшифровкой -  этого копи паста может быть достаточно.
</div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
credit_data = pd.read_csv('/datasets/data.csv')
credit_data.head(13)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   Среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
5          0    -926.185831         27    высшее             0   
6          0   -2879.202052         43    высшее             0   
7          0    -152.779569         50   СРЕДНЕЕ             1   
8          2   -6929.865299         35    ВЫСШЕЕ             0   
9          0   -2188.756445         41   среднее             1   
10         2   -4171.483647         36    высшее             0   
11         0    -792.701887         40   среднее             1   
12         0            NaN         65   среднее             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу

In [2]:
credit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод**

Изучив общую информацию статистики о платёжеспособности клиентов (*/data.csv*),  
Всего имеется 21525 строк с данными о заеме средств, из которых:  
1. В столбце **days_employed** выявлены:
    - отсутствующие значения;
    - неверный формат отображения данных(должен быть целочисленный);
    - наличие отрицательных значений;
    - имеются недопустимые данные - выбросы, например *id4* проработал 340266 дней, это составляет около 930 лет, что физически пока невозможно.
    
    Предположение: Если проанализировать общий датасет, то данные в столбце **total_income** отсутствуют в том же количестве, что и в исследуемом столбце, что говорит о пропуске типа "Неслучайные", неверный формат данных и отрицательные значения столбца говорят об ошибках в выгрузке данных. 
    
    
2. В столбце **education** данные представлены в разном регистре.  
    Предположение: Заполнение данных без стандартизации.      


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Первый взгляд на данные хороший.
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Посмотрим строки с пропусками,так как пропусков days_employed и total_income равное количество, то если хоть в одной строке есть оба пропуска сразу, то и в остальных строках пропуски будут одновременно и в одной строке.

In [3]:
days_employed_nan = credit_data[credit_data['days_employed'].isna() == True]
days_employed_nan

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Найдем долю отсутствующих значений от всех данных:

In [4]:
days_employed_nan_ratio = days_employed_nan['days_employed'].isna().sum() / credit_data.shape[0] 
'Доля отсутствующих значений - {:.1%}'.format(days_employed_nan_ratio)

'Доля отсутствующих значений - 10.1%'

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Так же можно узнать долю прописав `credit_data.isna().mean()` - так немного короче, возьми к себе в профессиональный запас.
</div>

In [5]:
credit_data.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

<div class="alert alert-info"> <b>Комментарий студента:</b> Ну это колдовство какое-то!!</div>

**Вывод**

10% это довольно много, тем более что по заданию кредитного отдела нам нужны другие метрики <code>&gt; влияет ли **семейное положение** и **количество детей** клиента на факт погашения кредита в срок.</code>, и если мы их удалим, потеряем часть важной информации, по этому заменим отсутствующие значения на средние по столбцу.

Но сначала применим фильтр:
- переведем значения в столбце **days_employed** в положительные;
- ячейки, в которых стаж выше чем 50 лет заменим на максимальный стаж(18250 дней);
- отсутствующие значения заменим на средние.

In [6]:
# max_days_employed = 18250 # 365*50=18250
# def days(day):
#     try:
#         if day < 0:
#             day = -day
#         if day > max_days_employed:
#             day = max_days_employed
#         return day
#     except: 'проверьте значения!'

# credit_data['days_employed'] = credit_data['days_employed'].apply(days)
# changing = credit_data[credit_data['days_employed'] == max_days_employed]['days_employed'].count()
# f'заменено {changing} значений'

In [7]:
max_days_employed = credit_data[credit_data['days_employed'] < 18250]['days_employed'].idxmax() #Заменил поиск максимума 
def days(day):
    try:
        if day < 0:
            day = -day
        if day > max_days_employed:
            day = max_days_employed
        return day
    except: 'проверьте значения!'

credit_data['days_employed'] = credit_data['days_employed'].apply(days)
changing = credit_data[credit_data['days_employed'] == max_days_employed]['days_employed'].count()
f'заменено {changing} значений, новый максимум: {max_days_employed}'
credit_data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   Среднее             1   
3             3    4124.747207         32   среднее             1   
4             0   17437.000000         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0   17437.000000         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Очень хорошо, молодец. А как думаешь с чем могут быть связаны такие аномальные значения?
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Мне кажется, просто кто-то либо неправильно свел таблицу, или просто ввел неверное значение.  
    А можно ли оставить выбросы, где количество дней заведомо большое, а при обработке упорядочить все значения по количеству дней и отрезать первые и последние 2-3% таблицы, оставив одну середину с более-менее корректными значениями, если так можно, то как это правильно сделать?(про исключение крайних значений)</div>

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

Если ты хочешь оценить выбросы, то для этого подойдут квантили из метода `describe` так же визуально их можно увидеть после использования боксплота. Изучи эти два метода и должно встать все на свои места.
</div>

In [8]:
# mean_days = credit_data['days_employed'].mean()
# mean_income = int(credit_data['total_income'].mean())
# credit_data['days_employed'] = credit_data['days_employed'].fillna(mean_days)
# credit_data['total_income'] = credit_data['total_income'].fillna(mean_income)
# mean_income

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Заполнять пропуски в `days_employed` и `total_income` просто средним вряд ли допустимо. Один отражает стаж, другой - доход. Оба параметра могут быть зависимы от других, например, стаж может зависеть от возраста, типа занятости, а доход, в свою очередь, может зависеть от образования и так же от типа занятости.
    
Дам рекомендации по заполнению пропусков:
    
- Оцени кол-во пропусков, много их или мало? При большом кол-ве заполнять их одним значением вряд ли хорошее решение, т.к. в данных есть и студенты, и пенсионеры, у них разное образование и т.п. Стаж так же не может быть одинаковым. Аналогично и с доходом, он так же много от чего зависит.
- Ты предобработал данные - супер, теперь, заполни пропуски исходя из разных типов занятости к примеру.
</div>

In [9]:
credit_data_grouped_income_mean = credit_data.groupby('income_type')['total_income'].mean()
credit_data_grouped_income_mean

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        170898.309923
компаньон          202417.461462
пенсионер          137127.465690
предприниматель    499163.144947
сотрудник          161380.260488
студент             98201.625314
Name: total_income, dtype: float64

In [10]:
credit_data_grouped_days_mean = credit_data.groupby('income_type')['days_employed'].mean()
credit_data_grouped_days_mean

income_type
безработный        17437.000000
в декрете           3296.759962
госслужащий         3399.896902
компаньон           2111.485385
пенсионер          17437.000000
предприниматель      520.848083
сотрудник           2326.404154
студент              578.751554
Name: days_employed, dtype: float64

In [11]:
credit_data['days_employed']= credit_data['days_employed'].fillna(0) 
credit_data['total_income']= credit_data['total_income'].fillna(0)

def func_income(row):
    income = row['total_income']
    inc_type = row['income_type']
    if income == 0: # if pd.isna(row['total_income']): почему-то не работает
        return credit_data_grouped_income_mean.loc[row['income_type']]
    return row['total_income']

def func_days(row):
    income = row['days_employed']
    inc_type = row['income_type']
    if income == 0: # if pd.isna(row['total_income']): почему-то не работает
        return credit_data_grouped_days_mean.loc[row['income_type']]
    return row['total_income']
 
credit_data['total_income'] = credit_data.apply(func_income, axis=1)
credit_data['days_employed'] = credit_data.apply(func_days, axis=1)

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

Теперь, хорошо.
</div>

In [12]:
credit_data.loc[11:13]

children  days_employed  dob_years            education  education_id  \
11         0   77069.234271         40              среднее             1   
12         0   17437.000000         65              среднее             1   
13         0  130458.228857         54  неоконченное высшее             2   

       family_status  family_status_id gender income_type  debt  \
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   

     total_income                            purpose  
11   77069.234271  покупка коммерческой недвижимости  
12  137127.465690                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля

Заменили отсутствующие значения средними по выборке из определенной категории `income_type`

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправил!</div>

Проверим остальные столбцы, в столбце **dob_years**:

In [13]:
years_min = credit_data['dob_years'].min()
years_max = credit_data['dob_years'].max()
f"минимальный возраст:{years_min}, максимальный возраст:{years_max}"


'минимальный возраст:0, максимальный возраст:75'

В возрасте 0 лет обычно не дают кредит, по-этому заменим на средние значения по группе `income_type`. `# Исправление после проверки` 

In [14]:
# years_median = credit_data['dob_years'].median()
# credit_data['dob_years'] = credit_data['dob_years'].replace(0,years_median)
# credit_data

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Заполнять 0 значения возраста медианой вряд ли допустимо. Данные разношерстные, есть разные возраста, разные типы занятости и у всех у них будет медианный возраст. Думаю, вряд ли такое встречается, чтобы студент и пенсионер были одного возраста?
    
- Попробуй заполнить 0 значения исходя из разных типов занятости к примеру
</div>

In [15]:
credit_data_grouped_years_mean = credit_data.groupby('income_type')['dob_years'].mean()
credit_data_grouped_years_mean

income_type
безработный        38.000000
в декрете          39.000000
госслужащий        40.636737
компаньон          39.697542
пенсионер          59.063019
предприниматель    42.500000
сотрудник          39.821027
студент            22.000000
Name: dob_years, dtype: float64

In [16]:
cd = credit_data[credit_data['dob_years'] == 0]
cd

children  days_employed  dob_years education  education_id  \
99            0   71291.522491          0   Среднее             1   
149           0   70176.435951          0   среднее             1   
270           3  102166.458894          0   среднее             1   
578           0   97620.687042          0   среднее             1   
1040          0  303994.134987          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0    2326.404154          0   среднее             1   
20462         0  259193.920299          0   среднее             1   
20577         0  129788.762899          0   среднее             1   
21179         2  240702.007382          0    высшее             0   
21313         0  152419.184851          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
149     70176.435951                       операции с жильем  
270    102166.458894                            ремонт жилью  
578     97620.687042  строительство собственной недвижимости  
1040   303994.134987                         свой автомобиль  
...              ...                                     ...  
19829  161380.260488                                   жилье  
20462  259193.920299                    покупка своего жилья  
20577  129788.762899                            недвижимость  
21179  240702.007382        строительство жилой недвижимости  
21313  152419.184851                       операции с жильем  

[101 rows x 12 columns]

In [17]:
def func_years(row):
    years = row['dob_years']
    inc_type = row['income_type']
    if years == 0: 
        return credit_data_grouped_years_mean.loc[row['income_type']]
    return row['dob_years']
credit_data['dob_years'] = credit_data.apply(func_years, axis=1)
credit_data.loc[148:150]

children  days_employed  dob_years education  education_id  \
148         0  145159.214977  37.000000   среднее             1   
149         0   70176.435951  39.821027   среднее             1   
150         0  174088.293221  71.000000   среднее             1   

        family_status  family_status_id gender income_type  debt  \
148   женат / замужем                 0      F   сотрудник     0   
149         в разводе                 3      F   сотрудник     0   
150  гражданский брак                 1      F   пенсионер     0   

      total_income                       purpose  
148  145159.214977  на покупку своего автомобиля  
149   70176.435951             операции с жильем  
150  174088.293221               сыграть свадьбу

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправил!</div>

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

Отлично. Так же необязательно заполнять их сначала нулями, можно сразу заменять пропущенные с помощью `fillna()`, только надо передать туда нужные значения.
</div>

Проверим столбец
**children**

In [18]:
credit_data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

выбросы: (-1) - заменим на 1, предполагая, что ошибка в выгрузке данных  
(20) - заменим на 2, предполагая, что ошибка в выгрузке данных

In [19]:
credit_data.loc[credit_data['children'] == -1,'children'] = 1
credit_data.loc[credit_data['children'] == 20,'children'] = 2
credit_data['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Да, тут согласен.
</div>

### Замена типа данных

Переведем значения столбцов **days_employed** и **total_income** в int, столбец **education** понизим в регистре.

In [20]:
credit_data['days_employed'] = credit_data['days_employed'].astype('int')
credit_data['total_income'] = credit_data['total_income'].astype('int')

In [21]:
credit_data['education'] = credit_data['education'].str.lower()

**Вывод**

Мы привели значения столбцов в int, понизили регистр для нахождения дубликатов

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Работа с типами корректна.
</div>

### Обработка дубликатов

In [22]:
credit_data.duplicated().sum()
credit_data.drop_duplicates().reset_index(drop=True)


children  days_employed  dob_years education  education_id  \
0             1         253875       42.0    высшее             0   
1             1         112080       36.0   среднее             1   
2             0         145885       33.0   среднее             1   
3             3         267628       32.0   среднее             1   
4             0         158616       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21449         1         224791       43.0   среднее             1   
21450         0         155999       67.0   среднее             1   
21451         1          89672       38.0   среднее             1   
21452         3         244093       38.0   среднее             1   
21453         2          82047       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21449  гражданский брак                 1      F   компаньон     0   
21450   женат / замужем                 0      F   пенсионер     0   
21451  гражданский брак                 1      M   сотрудник     1   
21452   женат / замужем                 0      M   сотрудник     1   
21453   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21449        224791             операции с жильем  
21450        155999          сделка с автомобилем  
21451         89672                  недвижимость  
21452        244093  на покупку своего автомобиля  
21453         82047         на покупку автомобиля  

[21454 rows x 12 columns]

**Вывод**

Выявлено 55 дублирующих строк.Они удалены методом drop_duplicates(), индексы удалены и заменены на новые.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Хорошо, согласен.
</div>

### Лемматизация

Проверим, на что же берут кредиты:

In [23]:
from pymystem3 import Mystem
from collections import Counter
m = Mystem()
purposes = credit_data['purpose'].unique()
purposes_len = len(purposes)
purposes, purposes_len




(array(['покупка жилья', 'приобретение автомобиля',
        'дополнительное образование', 'сыграть свадьбу',
        'операции с жильем', 'образование', 'на проведение свадьбы',
        'покупка жилья для семьи', 'покупка недвижимости',
        'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
        'строительство собственной недвижимости', 'недвижимость',
        'строительство недвижимости', 'на покупку подержанного автомобиля',
        'на покупку своего автомобиля',
        'операции с коммерческой недвижимостью',
        'строительство жилой недвижимости', 'жилье',
        'операции со своей недвижимостью', 'автомобили',
        'заняться образованием', 'сделка с подержанным автомобилем',
        'получение образования', 'автомобиль', 'свадьба',
        'получение дополнительного образования', 'покупка своего жилья',
        'операции с недвижимостью', 'получение высшего образования',
        'свой автомобиль', 'сделка с автомобилем',
        'профильное образов

Необходимо Лемматизировать список и попробовать еще раз

In [24]:
def lem(row):
    lemmas = m.lemmatize(row)
#     lemmas =' '.join(lemmas)
    return lemmas[:-1]
credit_data['purpose_lem'] = credit_data['purpose'].apply(lem)
credit_data

children  days_employed  dob_years education  education_id  \
0             1         253875       42.0    высшее             0   
1             1         112080       36.0   среднее             1   
2             0         145885       33.0   среднее             1   
3             3         267628       32.0   среднее             1   
4             0         158616       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1         224791       43.0   среднее             1   
21521         0         155999       67.0   среднее             1   
21522         1          89672       38.0   среднее             1   
21523         3         244093       38.0   среднее             1   
21524         2          82047       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  \
0            253875                 покупка жилья   
1            112080       приобретение автомобиля   
2            145885                 покупка жилья   
3            267628    дополнительное образование   
4            158616               сыграть свадьбу   
...             ...                           ...   
21520        224791             операции с жильем   
21521        155999          сделка с автомобилем   
21522         89672                  недвижимость   
21523        244093  на покупку своего автомобиля   
21524         82047         на покупку автомобиля   

                                    purpose_lem  
0                           [покупка,  , жилье]  
1                 [приобретение,  , автомобиль]  
2                           [покупка,  , жилье]  
3              [дополнительный,  , образование]  
4                         [сыграть,  , свадьба]  
...                                         ...  
21520                [операция,  , с,  , жилье]  
21521             [сделка,  , с,  , автомобиль]  
21522                            [недвижимость]  
21523  [на,  , покупка,  , свой,  , автомобиль]  
21524           [на,  , покупка,  , автомобиль]  

[21525 rows x 13 columns]

In [25]:
# purposes = credit_data['purpose_lem'].unique()
# purposes_len = len(purposes)
# purposes, purposes_len

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Выводи для начала частотность слов после лемматизации, чтобы твои категории явно выделились, для этого они должны быть в топе, иначе непонятно откуда они взялись.
</div>

In [26]:
from collections import Counter
super_vector = []
for purpose in credit_data['purpose_lem']:
    super_vector.extend(purpose)
Counter(super_vector)

Counter({'покупка': 5912,
         ' ': 33677,
         'жилье': 4473,
         'приобретение': 462,
         'автомобиль': 4315,
         'дополнительный': 909,
         'образование': 4022,
         'сыграть': 774,
         'свадьба': 2348,
         'операция': 2610,
         'с': 2924,
         'на': 2233,
         'проведение': 777,
         'для': 1294,
         'семья': 641,
         'недвижимость': 6367,
         'коммерческий': 1315,
         'жилой': 1233,
         'строительство': 1881,
         'собственный': 635,
         'подержать': 479,
         'свой': 2235,
         'со': 630,
         'заниматься': 908,
         'сделка': 944,
         'подержанный': 489,
         'получение': 1316,
         'высокий': 1375,
         'профильный': 436,
         'сдача': 653,
         'ремонт': 612})

<div class="alert alert-info"> <b>Комментарий студента: Исправил!</div>

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

Супер!
</div>

In [27]:
def in_purpose(purpose):
    if 'жилье' in purpose:
        return 'жилье'
    elif 'недвижимость' in purpose:
        return 'жилье'
    elif 'образование' in purpose:
        return 'образование'
    elif 'образование' in purpose:
        return 'образование'
    elif 'свадьба' in purpose:
        return 'свадьба'
    elif 'автомобиль' in purpose:
        return 'автомобиль'
    else: return 'nan'
    
credit_data['purpose_lem_stem'] = credit_data['purpose_lem'].apply(in_purpose)
credit_data

children  days_employed  dob_years education  education_id  \
0             1         253875       42.0    высшее             0   
1             1         112080       36.0   среднее             1   
2             0         145885       33.0   среднее             1   
3             3         267628       32.0   среднее             1   
4             0         158616       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1         224791       43.0   среднее             1   
21521         0         155999       67.0   среднее             1   
21522         1          89672       38.0   среднее             1   
21523         3         244093       38.0   среднее             1   
21524         2          82047       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  \
0            253875                 покупка жилья   
1            112080       приобретение автомобиля   
2            145885                 покупка жилья   
3            267628    дополнительное образование   
4            158616               сыграть свадьбу   
...             ...                           ...   
21520        224791             операции с жильем   
21521        155999          сделка с автомобилем   
21522         89672                  недвижимость   
21523        244093  на покупку своего автомобиля   
21524         82047         на покупку автомобиля   

                                    purpose_lem purpose_lem_stem  
0                           [покупка,  , жилье]            жилье  
1                 [приобретение,  , автомобиль]       автомобиль  
2                           [покупка,  , жилье]            жилье  
3              [дополнительный,  , образование]      образование  
4                         [сыграть,  , свадьба]          свадьба  
...                                         ...              ...  
21520                [операция,  , с,  , жилье]            жилье  
21521             [сделка,  , с,  , автомобиль]       автомобиль  
21522                            [недвижимость]            жилье  
21523  [на,  , покупка,  , свой,  , автомобиль]       автомобиль  
21524           [на,  , покупка,  , автомобиль]       автомобиль  

[21525 rows x 14 columns]

In [28]:
credit_data[credit_data['purpose_lem_stem'] == 'nan']['purpose_lem_stem'].count()

0

**Вывод**

Мы определили список уникальных целей кредита, выделили основные слова и промаркировали их. Отсутствующих значений нет.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

C  точки зрения кода лемматизация корректна, стоит только поработать над выводом частотности.
</div>

### Категоризация данных

Определим, кто имел долги  по кредитам и попробуем разобраться в чем причина. Начнем с зависимости количества детей от наличия задолжности, составим идентификатор: есть дети:1, нет детей:0

In [29]:
def children_id_f (row):
    if row > 0:
        return 1
    else: return 0
credit_data['children_id'] = credit_data['children'].apply(children_id_f)
child_debt = credit_data.groupby('children_id')['debt'].count()
child_debt

children_id
0    14149
1     7376
Name: debt, dtype: int64

Вывод: У кого нет детей, в 2 раза чаще берут кредит!  
Определим, какую часть от обеих групп составляют люди, не вернувшие долг и сравним их.
Доля не вернувших кредит из числа людей, не имеющих детей:

In [30]:
no_child_bad_debt = credit_data[(credit_data['children_id'] == 0) & (credit_data['debt'] == 1)]['children_id'].count()
no_child_bad_debt

1063

In [31]:
no_child_ratio = no_child_bad_debt / child_debt[0]
'Доля не вернувших кредит из числа людей, не имеющих детей:{:.1%}'.format(no_child_ratio)

'Доля не вернувших кредит из числа людей, не имеющих детей:7.5%'

Доля не вернувших кредит из числа людей, имеющих 1 и более детей:

In [32]:
child_bad_debt = credit_data[(credit_data['children_id'] == 1) & (credit_data['debt'] == 1)]['children_id'].count()
child_bad_debt

678

In [33]:
child_ratio = child_bad_debt / child_debt[1]
'Доля не вернувших кредит из числа людей, имеющих 1 и более детей:{:.1%}'.format(child_ratio)

'Доля не вернувших кредит из числа людей, имеющих 1 и более детей:9.2%'

**Вывод**

Мы провели категоризацию данных по количеству невозвращенных кредитов и количеству детей у заемщикаЮ определили вероятность возвращения кредита от количества детей.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Супер, вывод корректный. Рекомендую пользоваться группировкой как при ответе на вопрос с влиянием `family_status_id` на задолжность.
</div>

Проверим зависимость семейного положения от задолжностей.

In [34]:
credit_data_fam_table = credit_data[['family_status', 'family_status_id']]
d = credit_data_fam_table.drop_duplicates()
d


family_status  family_status_id
0         женат / замужем                 0
4        гражданский брак                 1
18         вдовец / вдова                 2
19              в разводе                 3
24  Не женат / не замужем                 4

In [35]:
credit_data_fam = credit_data.groupby('family_status_id')['debt'].count()
credit_data_fam

family_status_id
0    12380
1     4177
2      960
3     1195
4     2813
Name: debt, dtype: int64

Женатые/замужем чаще всего берут кредиты, найдем тех, кто чаще всего не возвращает. так как в столбце **debt** имеющие задолжности маркируются как "1", то сумма всех значения покажет количество людей, имевших задолжность.

In [36]:
credit_data_fam_debt = credit_data.groupby('family_status_id')['debt'].sum()
credit_data_fam_debt

family_status_id
0    931
1    388
2     63
3     85
4    274
Name: debt, dtype: int64

In [37]:
с = credit_data_fam_debt / credit_data_fam
с*100

family_status_id
0    7.520194
1    9.288963
2    6.562500
3    7.112971
4    9.740491
Name: debt, dtype: float64

9.7% людей со статусом Не женат / не замужем имели задолжности

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Отлично, тут все корректно, дам рекомендацию, попробуй изменить в группировке `sum` на `mean`, тогда действий станет меньше.
</div>

Проверим зависимость между уровнем дохода и возвратом кредита в срок.

In [38]:
mean_income = int(credit_data['total_income'].mean())
credit_data_income_debt = credit_data[['total_income', 'debt']]
credit_data_income_debt
def income(value):
    if value < mean_income:
        return 'меньше среднего'
    else: return 'больше среднего'
credit_data['total_income_value'] = credit_data['total_income'].apply(income)
credit_data


children  days_employed  dob_years education  education_id  \
0             1         253875       42.0    высшее             0   
1             1         112080       36.0   среднее             1   
2             0         145885       33.0   среднее             1   
3             3         267628       32.0   среднее             1   
4             0         158616       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1         224791       43.0   среднее             1   
21521         0         155999       67.0   среднее             1   
21522         1          89672       38.0   среднее             1   
21523         3         244093       38.0   среднее             1   
21524         2          82047       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  \
0            253875                 покупка жилья   
1            112080       приобретение автомобиля   
2            145885                 покупка жилья   
3            267628    дополнительное образование   
4            158616               сыграть свадьбу   
...             ...                           ...   
21520        224791             операции с жильем   
21521        155999          сделка с автомобилем   
21522         89672                  недвижимость   
21523        244093  на покупку своего автомобиля   
21524         82047         на покупку автомобиля   

                                    purpose_lem purpose_lem_stem  children_id  \
0                           [покупка,  , жилье]            жилье            1   
1                 [приобретение,  , автомобиль]       автомобиль            1   
2                           [покупка,  , жилье]            жилье            0   
3              [дополнительный,  , образование]      образование            1   
4                         [сыграть,  , свадьба]          свадьба            0   
...                                         ...              ...          ...   
21520                [операция,  , с,  , жилье]            жилье            1   
21521             [сделка,  , с,  , автомобиль]       автомобиль            0   
21522                            [недвижимость]            жилье            1   
21523  [на,  , покупка,  , свой,  , автомобиль]       автомобиль            1   
21524           [на,  , покупка,  , автомобиль]       автомобиль            1   

      total_income_value  
0        больше среднего  
1        меньше среднего  
2        меньше среднего  
3        больше среднего  
4        меньше среднего  
...                  ...  
21520    больше среднего  
21521    меньше среднего  
21522    меньше среднего  
21523    больше среднего  
21524    меньше среднего  

[21525 rows x 16 columns]

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

С категоризацией согласен, идем дальше.
</div>

In [39]:
# income_value_debt = credit_data.groupby('total_income_value')['debt'].sum()
# more_mean_income_value = income_value_debt[0] / credit_data.shape[0]
# less_mean_income_value = income_value_debt[1] / credit_data.shape[0]
# 'Люди, имеющие доход, меньше среднего по выборке(167 422 руб) имели задолжность в {:.1%}, больше среднего - в {:.1%}'.format(more_mean_income_value,less_mean_income_value)

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>

Если я правильно тебя понял, то Люди, имеющие доход, меньше среднего по выборке(167 422 руб) имели задолжность в 3.4% случаев, а больше среднего - в 4.7% случаев. Для поиска доли должников нужно разделить кол-во должников на `общее кол-во людей в этой категории`. При ответах на другие вопросы расчеты были корректными.
</div>

In [40]:
income_value_debt_negative = credit_data.groupby('total_income_value')['debt'].sum()
income_value_debt = credit_data.groupby('total_income_value')['debt'].count()
more_mean_income_value = income_value_debt_negative[0] / income_value_debt[0]
less_mean_income_value = income_value_debt_negative[1] / income_value_debt[1]
'Люди, имеющие доход, меньше среднего по выборке(167 422 руб) имели задолжность в {:.1%}, больше среднего - в {:.1%}'.format(more_mean_income_value,less_mean_income_value)

'Люди, имеющие доход, меньше среднего по выборке(167 422 руб) имели задолжность в 7.5%, больше среднего - в 8.5%'

<div class="alert alert-info"> <b>Комментарий студента: Исправил!</div>

<div class="alert alert-success">
<b>Комментарий ревьюера V2:</b>

Отлично.
</div>

Люди,имеющие доход больше среднего чаще всего имеют задолжности

Проверим зависимость между целью кредита и возвратом кредита в срок.

In [41]:
purpose_debt_value = credit_data['purpose_lem_stem'].value_counts()
purpose_debt_value

жилье          10840
автомобиль      4315
образование     4022
свадьба         2348
Name: purpose_lem_stem, dtype: int64

На жилье чаще всего берут кредиты!

In [42]:
purpose_debt = credit_data.groupby('purpose_lem_stem')['debt'].sum().sort_values(ascending=False)
purpose_debt

purpose_lem_stem
жилье          782
автомобиль     403
образование    370
свадьба        186
Name: debt, dtype: int64

Также на жилье чаще всего и не возвращают заем.

In [43]:
purpose_debt_ratio = purpose_debt / purpose_debt_value
purpose_debt_ratio*100

purpose_lem_stem
жилье          7.214022
автомобиль     9.339513
образование    9.199403
свадьба        7.921635
dtype: float64

Но в процентном соотношении, чаще всего не возвращаются кредиты на автомобиль.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

А вот здесь у тебя посчитано все корректно, доли выделены верно.
</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

**Вывод**

Люди, у которых нет детей чаще возвращают долг, чем люди, у которых есть дети.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Вывод**

Люди со статусом Не женат / не замужем имели задолжности чаще других.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Вывод**

Люди,имеющие доход больше среднего чаще всего имеют задолжности

- Как разные цели кредита влияют на его возврат в срок?

**Вывод**

Чаще всего не возвращаются кредиты на автомобиль.

## Шаг 4. Общий вывод

Получена таблица из 21525 строк, в котором были обнаружены пропущенные значения, которые были заменены на средние по столбцу, дубликаты, которые были удалены из таблицы, выбросы, которые были обработаны в частном порядке. Данные разбивались по категриям и исследовались. Из исследования следует вывод, что люди, у которых есть дети, с семейным положением не женат / не замужем, с доходом выше среднего и с целью кредита на автомобиль чаще всего не возвращают кредит. 

In [44]:
credit_data[(credit_data['children_id'] == 1)&(credit_data['family_status_id'] == 0)&(credit_data['total_income'] > mean_income)&(credit_data['purpose_lem_stem'] == 'автомобиль')].head()


children  days_employed  dob_years education  education_id  \
46          2         182682       37.0   среднее             1   
93          1         331139       32.0   среднее             1   
132         1         246227       45.0   среднее             1   
164         2         369587       27.0    высшее             0   
177         2         173074       48.0   среднее             1   

       family_status  family_status_id gender  income_type  debt  \
46   женат / замужем                 0      F    компаньон     0   
93   женат / замужем                 0      M    сотрудник     1   
132  женат / замужем                 0      M    сотрудник     0   
164  женат / замужем                 0      F    сотрудник     1   
177  женат / замужем                 0      M  госслужащий     0   

     total_income                           purpose  \
46         182682  сделка с подержанным автомобилем   
93         331139              сделка с автомобилем   
132        246227                        автомобиль   
164        369587      на покупку своего автомобиля   
177        173074  сделка с подержанным автомобилем   

                                       purpose_lem purpose_lem_stem  \
46   [сделка,  , с,  , подержанный,  , автомобиль]       автомобиль   
93                   [сделка,  , с,  , автомобиль]       автомобиль   
132                                   [автомобиль]       автомобиль   
164       [на,  , покупка,  , свой,  , автомобиль]       автомобиль   
177  [сделка,  , с,  , подержанный,  , автомобиль]       автомобиль   

     children_id total_income_value  
46             1    больше среднего  
93             1    больше среднего  
132            1    больше среднего  
164            1    больше среднего  
177            1    больше среднего

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Общий вывод принят.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.